query/src/main/java/org/infinispan/search/mapper/work/impl/SearchIndexerImpl.java

```diff
@@ -1,39 +1,112 @@
+import org.infinispan.query.logging.Log;
 import org.infinispan.search.mapper.mapping.EntityConverter;
 import org.infinispan.search.mapper.session.impl.InfinispanIndexedTypeContext;
 import org.infinispan.search.mapper.session.impl.InfinispanTypeContextProvider;
 import org.infinispan.search.mapper.work.SearchIndexer;
-import org.infinispan.util.concurrent.BlockingManager;
+import org.infinispan.util.logging.LogFactory;
+
+import io.reactivex.rxjava3.core.Flowable;
+import io.reactivex.rxjava3.disposables.Disposable;
+import io.reactivex.rxjava3.internal.functions.Functions;
+import io.reactivex.rxjava3.processors.FlowableProcessor;
+import io.reactivex.rxjava3.processors.UnicastProcessor;
 
 /**
  * @author Fabio Massimo Ercoli
  */
 public class SearchIndexerImpl implements SearchIndexer {
+   private static final Log log = LogFactory.getLog(QueryInterceptor.class, Log.class);
 
    private final PojoIndexer delegate;
    private final EntityConverter entityConverter;
    private final InfinispanTypeContextProvider typeContextProvider;
 
-   private final SimpleScheduledExecutor offloadingExecutor;
+   private final int maxConcurrency;
+   private final ConcurrentMap<Supplier<Flowable<?>>, CompletableFuture<?>> submittedTasks;
+
+   private Disposable processorDisposer;
+
+   private FlowableProcessor<Supplier<Flowable<?>>> requestProcessor;
 
    public SearchIndexerImpl(PojoIndexer delegate, EntityConverter entityConverter,
-         InfinispanTypeContextProvider typeContextProvider, BlockingManager blockingManager) {
+                            InfinispanTypeContextProvider typeContextProvider, int maxConcurrency) {
       this.delegate = delegate;
       this.entityConverter = entityConverter;
       this.typeContextProvider = typeContextProvider;
-      this.offloadingExecutor = InfinispanIndexingExecutorProvider.writeExecutor(blockingManager);
+      this.maxConcurrency = maxConcurrency;
+
+      this.submittedTasks = new ConcurrentHashMap<>();
+   }
+
+   @Override
+   public void start() {
+      requestProcessor = UnicastProcessor.<Supplier<Flowable<?>>>create().toSerialized();
+      // onBackpressureLatest will drop any item that it can't immediately pass downstream
+      // Thus next downstream must batch them to not immediately drop
+      processorDisposer = requestProcessor.onBackpressureLatest(supplier -> {
+               CompletableFuture<?> completableFuture = submittedTasks.remove(supplier);
+               if (completableFuture == null) {
+                  throw new IllegalStateException("Dropped task " + supplier + " not found in submittedTask " + submittedTasks);
+               }
+               completableFuture.completeExceptionally(log.hibernateSearchBackpressure());
+            })
+            // We give HS some extra buffer room before we start throwing back pressure exceptions if
+            // the flatMap operations can't keep up
+            .rebatchRequests(10_000)
```

Comment:
Should it be some % over the `maxConcurrency`? It is 10x default size, which I guess should be more than enough.

Reply:
I don't think so. The goal here is just a buffer of some sort. I doubt 10K is going to break the bank, but also give us some buffer room before throwing exceptions when the user doesn't change their index write queue sizes.

For example our IndexingOperationOffloadingTest sets the queue size to 1, which means we would only have a buffer of 1 here and any write concurrency over 2 would cause failures immediately.

But I am more than open to having this changed at some point. The final goal would be to lower this number and instead of immediately throwing exceptions instead to notify some part of the system "Hey we need to slow operations" and if that doesn't stop subsequent pressure then throw the exceptions.